### PR TITLE
Add spin bit (without VEC)

### PR DESCRIPTION
**Pull request by britram**
`cmake -DSPINBIT=1` to enable

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 .DS_Store
 .vagrant/
+build/
```
Comment (larseggert) [inline .gitignore]: Remove this

File: lib/src/conn.h [hunk 2]
```diff
@@ -126,7 +126,12 @@ struct q_conn {
     uint16_t err_code;
     uint8_t err_frm;
 
+#ifdef SPINBIT
+    uint8_t next_spin : 1;
```
Comment (larseggert) [inline lib/src/conn.h]: Make this part of the existing options bitmask?

File: lib/src/pkt.c [hunk 3]
```diff
@@ -234,9 +234,22 @@ bool enc_pkt(struct q_stream * const s,
...
 
-    if (is_set(F_LONG_HDR, meta(v).hdr.flags) == false)
+    if (is_set(F_LONG_HDR, meta(v).hdr.flags) == false) {
+#ifdef SPINBIT
+        // clear spin/vec
+        meta(v).hdr.flags &= ~F_SH_EXP_MASK;
+
+        // set spin bit
+        if (c->next_spin) {
+            meta(v).hdr.flags |= F_SH_SPIN;
```
Comment (larseggert) [inline lib/src/pkt.c]: I don't do braces on single-line conditionals

File: lib/src/pkt.c [hunk 4]
```diff
@@ -234,9 +234,22 @@ bool enc_pkt(struct q_stream * const s,
...
+        }
+
+        warn(DBG, "setting spin bits to %x", meta(v).hdr.flags & F_SH_EXP_MASK);
```
Comment (larseggert) [inline lib/src/pkt.c]: 0x%02x

File: lib/src/pkt.c [hunk 5]
```diff
@@ -633,6 +646,21 @@ bool dec_pkt_hdr_remainder(struct w_iov * const v,
...
+        if (nr > diet_max(&(c->pn_data.pn.recv_all))) {
+            if (c->is_clnt) {
+                c->next_spin = ((meta(v).hdr.flags & F_SH_SPIN) == 0);
+                warn(DBG, "inverting spin to %x", c->next_spin);
+            } else {
+                c->next_spin = ((meta(v).hdr.flags & F_SH_SPIN) != 0);
+                warn(DBG, "reflecting spin to %x", c->next_spin);
+            }
+        } else {
+            warn(DBG, "not updating next_spin: %llx <= %llx", nr, diet_max(&(c->pn_data.pn.recv_all)));
```
Comment (larseggert) [inline lib/src/pkt.c]: Single-line conditional

File: lib/src/pkt.h [hunk 6]
```diff
@@ -47,6 +47,9 @@
 
 // #define F_SH_KYPH 0x40
 #define F_SH_EXP_MASK 0x07
+#define F_SH_SPIN 0x04
+#define F_SH_VEC1 0x02
```
Comment (larseggert) [inline lib/src/pkt.h]: Since the VEC bits aren't used, no need to define them

File: lib/src/conn.h (hunk 2 above)
Comment (britram) [reply lib/src/conn.h]: IIRC that's a uint16 with no space left in it, will check again.

Comment (britram) [reply lib/src/conn.h]: ...ah now it's a 32. OK, will do.

**Comment by larseggert** [comment]
I pushed some minor cleanup, please verify things are still spinning?

**Comment by britram** [comment]
yep, still works. thanks!

**Pull request merged**